packages/next/src/client/components/nav-failure-handler.ts

```diff
@@ -0,0 +1,39 @@
+import { useEffect } from 'react'
+import { createHrefFromUrl } from './router-reducer/create-href-from-url'
+
+export function useNavFailureHandler() {
+  if (process.env.__NEXT_APP_NAV_FAIL_HANDLING) {
+    // this if is only for DCE of the feature flag not conditional
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useEffect(() => {
+      const uncaughtExceptionHandler = (
+        evt: ErrorEvent | PromiseRejectionEvent
+      ) => {
+        const error = 'reason' in evt ? evt.reason : evt.error
+        // if we have an unhandled exception/rejection during
+        // a navigation we fall back to a hard navigation to
+        // attempt recovering to a good state
+        if (
+          window.next.__pendingUrl &&
+          createHrefFromUrl(new URL(window.location.href)) !==
+            createHrefFromUrl(window.next.__pendingUrl)
+        ) {
+          console.error(
+            `Error occurred during navigation, falling back to hard navigation`,
+            error
+          )
+          window.location.href = window.next.__pendingUrl.toString()
```

Comment:
up to you but we could probably move this whole `window.location.href` bit into a function that the error boundary could leverage as well so we don't need to duplicate the handling